core/tee/tee_fs_common.c

```diff
@@ -1128,73 +1143,124 @@ tee_fs_off_t tee_fs_common_lseek(struct tee_fs_fd *fdp,
 }
 
 /*
- * To ensure atomic ftruncate operation, we can:
+ * To ensure atomic truncate operation, we can:
  *
  *  - update file length to new length
  *  - commit new meta
+ *  - free unused blocks
  *
- * (Any failure in above steps is considered as update failed,
- *  and the file content will not be updated)
+ * To ensure atomic extend operation, we can:
  *
- * After previous step the update is considered complete, but
- * we should do the following clean-up step(s):
+ *  - update file length to new length
+ *  - allocate and fill zero data to new blocks
+ *  - commit new meta
  *
- *  - free the trucated file blocks
+ * Any failure before committing new meta is considered as
+ * update failed, and the file content will not be updated
  *
- * (Any failure in above steps is considered as a successfully
- *  update)
  */
 int tee_fs_common_ftruncate(struct tee_fs_fd *fdp,
 				tee_fs_off_t new_file_len)
 {
 	int res = -1;
-	size_t old_file_len;
-	int old_block_num, new_block_num, diff_blocks, i = 0;
+	size_t old_file_len = fdp->meta->info.length;
 	struct tee_fs_file_meta *new_meta = NULL;
+	uint8_t *buf = NULL;
 
 	if (!fdp)
-		goto exit;
+		return -1;
 
-	if (fdp->flags & TEE_FS_O_RDONLY)
-		goto exit;
+	if (fdp->flags & TEE_FS_O_RDONLY) {
+		EMSG("Read only");
```

Comment:
The error message would looks like:
ERR TEE-CORE:tee_fs_common_ftruncate:1174: Read only